cartography/intel/semgrep/findings.py

```diff
@@ -0,0 +1,217 @@
+import logging
+from typing import Any
+from typing import Dict
+from typing import List
+
+import neo4j
+import requests
+
+from cartography.client.core.tx import load
+from cartography.graph.job import GraphJob
+from cartography.models.semgrep.findings import SemgrepDeploymentSchema
+from cartography.models.semgrep.findings import SemgrepSCAFindingSchema
+from cartography.models.semgrep.findings import SemgrepSCALocationSchema
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+
+
+@timeit
+def get_deployment(semgrep_app_token: str) -> List[Dict[str, Any]]:
+    """
+    Gets the deployment associated with the passed Semgrep App token.
+    param: semgrep_app_token: The Semgrep App token to use for authentication.
+    """
+    deployment = {}
+    deployment_url = "https://semgrep.dev/api/v1/deployments"
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": f"Bearer {semgrep_app_token}",
+    }
+    response = requests.get(deployment_url, headers=headers)
+    response.raise_for_status()
+    try:
+        if response.status_code == 200:
+            data = response.json()
+            deployment["id"] = data["deployments"][0]["id"]
+            deployment["name"] = data["deployments"][0]["name"]
+            deployment["slug"] = data["deployments"][0]["slug"]
+    except requests.RequestException as e:
+        logger.error(
+            "Could not complete request to the deployments Semgrep API: %s",
+            e,
+            exc_info=e,
+        )
+        raise e
+    except Exception as e:
+        logger.error("Erorr retrieving Semgrep deployment info: %s", e, exc_info=e)
+        raise e
+    return [deployment]
+
+
+@timeit
+def get_sca_vulns(semgrep_app_token: str, deployment_id: str) -> List[Dict[str, Any]]:
+    """
+    Gets the SCA vulns associated with the passed Semgrep App token and deployment id.
+    param: semgrep_app_token: The Semgrep App token to use for authentication.
+    param: deployment_id: The Semgrep deployment id to use for retrieving SCA vulns.
+    """
+    all_vulns = []
+    sca_url = f"https://semgrep.dev/api/sca/deployments/{deployment_id}/vulns"
+    has_more = True
+    cursor = ""
+    prev_request = None
+    headers = {
+        "Content-Type": "application/json",
+        "Authorization": f"Bearer {semgrep_app_token}",
+    }
+
+    while has_more:
+        params = {}
+        if cursor:
+            params = {"cursor": cursor}
+        try:
+            response = requests.get(sca_url, params=params, headers=headers)
+            response.raise_for_status()
+            if prev_request == response.request.url:
+                logger.warning(
+                    "Duplicate request detected. Breaking the loop to avoid infinite requests.",
+                )
```

Comment:
I'm guessing this check was needed during initial development, but the situation shouldn't happen anymore with the way you've structured the loop?

Reply:
It was a safe guard, as nothing avoids to call `https://semgrep.dev/api/sca/deployments/{deployment_id}/vulns?cursor=lastcursor`.
Nevertheless, tested retrieving all the vulns without it and it worked. Removing it.